Mathlib/CategoryTheory/WithTerminal.lean

```diff
@@ -138,6 +141,116 @@ def map {D : Type*} [Category D] (F : C ⥤ D) : WithTerminal C ⥤ WithTerminal
     | star, star, _ => PUnit.unit
 #align category_theory.with_terminal.map CategoryTheory.WithTerminal.map
 
+/--A natural isomorphism between the functor `map (𝟭 C)` and `𝟭 (WithTerminal C)`.-/
+def mapId  (C : Type*) [Category C] : map (𝟭 C) ≅ 𝟭 (WithTerminal C) where
+  hom := {app  := fun X =>  eqToHom (by cases X <;> rfl)}
+  inv := {app  := fun X =>  eqToHom (by cases X <;> rfl)}
+
+/--A natural isomorphism between the functor `map (F⋙G) ` and `map F ⋙ map G `.-/
+def mapComp {D : Type*} [Category D] {E : Type*} [Category E] (F : C⥤ D) (G:D⥤ E) :
+    map (F⋙G) ≅ map F ⋙ map G where
+  hom := {app := fun X =>  eqToHom (by cases X <;> rfl)}
+  inv := {app  := fun X =>  eqToHom (by cases X <;> rfl)}
+
+/--From a natrual transformation of functors `C⥤D`, the induced natural transformation
+of functors `WithTerminal C ⥤ WithTerminal D`. -/
+def map₂  {D : Type*} [Category D]  {F G: C ⥤ D} (η : F ⟶ G) :
+    map F ⟶ map G where
+  app := fun X =>
+        match X with
+        | of x => η.app x
+        | star => 𝟙 (star)
+  naturality := by
+          intro X Y f
+          match X, Y, f with
+          | of x, of y, f => exact η.naturality f
+          | of x, star, _ => rfl
+          | star, star, _ => rfl
+
+/-- The pseudofunctor from `Cat` to `Cat` defined with `WithTerminal`.-/
+def pseudofunctor: Pseudofunctor Cat Cat where
+  obj C := Cat.of (WithTerminal C)
+  map {C D} F := map F
+  map₂ := map₂
+  mapId C := mapId C
+  mapComp {C D E} F G  := mapComp F G
+  map₂_id := by
+    intros
+    apply NatTrans.ext
+    funext X
+    cases X <;> rfl
+  map₂_comp := by
+    intros
+    apply NatTrans.ext
+    funext X
+    cases X <;> rfl
+  map₂_whisker_left := by
+    intros
+    apply NatTrans.ext
+    funext X
+    cases X
+    · rw [NatTrans.comp_app, NatTrans.comp_app]
+      unfold mapComp map  map₂
```

Comment:
The `unfold` tactic is very rarely used in mathlib. We always prefer to have good API lemmas instead of unfolding a definition deep inside a proof. Luckily, these lemmas can be autogenerated by the `simps` attribute. I would suggest tagging `map`, `mapComp` and `map₂` with `@[simps]` (or `@[simps!]` if necessary), then the simplifier should be able to unfold the definition in the right places. If you get problems with match statements appearing in your goals, then you might need to add some lemmas by hand along the lines of
```lean
@[simp]
lemma mapId_hom_app_of (X : C) : (mapId C).hom.app (of X) = 𝟙 (of X) :=
  rfl
```